xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,218 @@
+      Map<String, Expr> conditions = new LinkedHashMap<>();
+      for (Map.Entry<String, Policy> rbacPolicy: rbac.getPolicies().entrySet()) {
+        conditions.put(rbacPolicy.getKey(), rbacPolicy.getValue().getCondition());
+      }
+      if (rbac.getAction() == Action.ALLOW) {
+        allowEngine = new RbacEngine(Action.ALLOW, ImmutableMap.copyOf(conditions));
+      }
+      if (rbac.getAction() == Action.DENY) {
+        denyEngine = new RbacEngine(Action.DENY, ImmutableMap.copyOf(conditions));
+      }
+    }
+    this.allowEngine = allowEngine;
+    this.denyEngine = denyEngine;
+  }
+
+  /**
+   * The evaluate function performs the core authorization mechanism of CEL Evaluation Engine.
+   * Determines whether a gRPC call is allowed, denied, or unable to decide.
+   * @param args evaluate argument that is used to evaluate the RBAC conditions.
+   * @return an AuthorizationDecision generated by CEL Evaluation Engine.
+   */
+  public AuthorizationDecision evaluate(EvaluateArgs<ReqT, RespT> args) {
+    List<String> unknownPolicyNames = new ArrayList<>();
+    // Set up activation used in CEL library's eval function.
+    Activation activation = Activation.copyOf(extractFields(args));
+    // Iterate through denyEngine's map.
+    // If there is match, immediately return deny. 
+    // If there are unknown results, return undecided. 
+    // If all non-match, then iterate through allowEngine.
+    if (denyEngine != null) {
+      for (Map.Entry<String, Expr> condition : denyEngine.conditions.entrySet()) {
+        try {
+          if (matches(condition.getValue(), activation)) {
+            return new AuthorizationDecision(AuthorizationDecision.Decision.DENY, 
+                new ArrayList<String>(Arrays.asList(new String[] {condition.getKey()})));
+          }
+        } catch (InterpreterException e) {
+          unknownPolicyNames.add(condition.getKey());
+        }
+      }
+      if (unknownPolicyNames.size() > 0) {
+        return new AuthorizationDecision(
+            AuthorizationDecision.Decision.UNKNOWN, unknownPolicyNames);
+      }
+    }
+    // Once we enter allowEngine, if there is a match, immediately return allow. 
+    // In the end of iteration, if there are unknown rules, return undecided.
+    // If all non-match, return deny.
+    if (allowEngine != null) {
+      for (Map.Entry<String, Expr> condition : allowEngine.conditions.entrySet()) {
+        try {
+          if (matches(condition.getValue(), activation)) {
+            return new AuthorizationDecision(AuthorizationDecision.Decision.ALLOW, 
+                new ArrayList<String>(Arrays.asList(new String[] {condition.getKey()})));
+          }
+        } catch (InterpreterException e) {
+          unknownPolicyNames.add(condition.getKey());
+        }
+      }
+      if (unknownPolicyNames.size() > 0) {
+        return new AuthorizationDecision(
+            AuthorizationDecision.Decision.UNKNOWN, unknownPolicyNames);
+      }
+    }
```

Comment:
So the above 2 `if` clauses can be refactored to avoid repetition, right? The method takes an entrySet and AuthorizationDecision.Decision as params and you call it twice with (allowEngine.conditions.entrySet(), AuthorizationDecision.Decision.ALLOW) and (denyEngine.conditions.entrySet(), AuthorizationDecision.Decision.DENY).

Reply:
Yes! I'll change it to avoid duplicate code :)